metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -99,6 +100,36 @@ void DBGSSHash::add_sequence(std::string_view sequence,
     throw std::logic_error("adding sequences not supported");
 }
 
+void DBGSSHash
+::map_to_nodes_with_rc_advanced(std::string_view sequence,
+                                const std::function<void(sshash::lookup_result)>& callback,
+                                bool with_rc,
+                                const std::function<bool()>& terminate) const {
+    if (terminate() || sequence.size() < k_)
+        return;
+
+    std::visit([&](const auto &dict) {
+        using kmer_t = get_kmer_t<decltype(dict)>;
+
+        std::vector<char> seq_encoded;
+        seq_encoded.reserve(sequence.size());
+        for (size_t i = 0; i < sequence.size(); ++i) {
+            seq_encoded.emplace_back(!kmer_t::is_valid(sequence[i]));
+        }
+
+        auto invalid = utils::drag_and_mark_segments(seq_encoded, 1, k_);
+
+        kmer_t uint_kmer = sshash::util::string_to_uint_kmer<kmer_t>(sequence.data(), k_ - 1);
+        uint_kmer.pad_char();
+        for (size_t i = k_ - 1; i < sequence.size() && !terminate(); ++i) {
+            uint_kmer.drop_char();
+            uint_kmer.kth_char_or(k_ - 1, kmer_t::char_to_uint(sequence[i]));
+            callback(!invalid[i] ? dict.lookup_advanced_uint(uint_kmer, with_rc)
+                                 : sshash::lookup_result());
```

Comment:
`utils::drag_and_mark_segments` doesn't work with `std::vector<bool>` since we don't have direct access to its underlying storage, so I revised this slightly

Reply:
Do we actually need to have access to its internal storage for `drag_and_mark_segments`? Maybe a better thing to do would be to change https://github.com/ratschlab/metagraph/blob/f2f320fa3fd3c9472260abc2abe0857618d8181e/metagraph/src/common/algorithms.hpp#L56-L57

into

```cpp
 = std::find(array.begin(), array.end(), label) 
	     - array.begin();
```

Reply:
good point, I actually don't think this needs to use `data`, so we can just use iterators